Sources/NostrSDK/RelayPool.swift

```diff
@@ -0,0 +1,152 @@
+                  delegate: delegate)
+    }
+    
+    private func setUpRelays() {
+        relays.forEach { setUpRelay($0) }
+        updateEventsPublisher()
+    }
+    
+    private func setUpRelay(_ relay: Relay) {
+        relay.delegate = self
+        relay.connect()
+    }
+    
+    private func updateEventsPublisher() {
+        let mergedEvents = Publishers.MergeMany(relays.map { $0.events })
+        cancellable = mergedEvents
+            .sink { [weak self] event in
+                self?.events.send(event)
+            }
+    }
+    
+    /// Adds a relay to the pool, if a matching one is not already there (based on relay URL).
+    /// - Parameter relay: The relay to add to the pool.
+    ///
+    /// > Note: The relay will be automatically connected.
+    public func add(relay: Relay) {
+        guard !relays.contains(relay) else {
+            logger.warning("Could not add relay because it was already in the pool: \(relay.url)")
+            return
+        }
+        
+        relays.insert(relay)
+        setUpRelay(relay)
+        updateEventsPublisher()
+    }
+    
+    /// Removes a relay from the pool, if a matching one is found (based on relay URL).
+    /// - Parameter relay: The relay to remove from the pool.
+    ///
+    /// > Note: The relay connection will be automatically closed.
+    public func remove(relay: Relay) {
+        removeRelay(withURL: relay.url)
+        updateEventsPublisher()
+    }
+    
+    /// Removes a relay from the pool, if one with a matching URL is found.
+    /// - Parameter relayURL: The relay URL to remove.
+    ///
+    /// > Note: The relay connection will be automatically closed.
+    public func removeRelay(withURL relayURL: URL) {
+        let matches = relays.filter { $0.url == relayURL }
+        matches.forEach {
+            $0.disconnect()
+            relays.remove($0)
+        }
+    }
+    
+    // MARK: - RelayOperating
+    
+    /// Attempts to connect to all of the relays.
+    public func connect() {
+        relays.forEach { $0.connect() }
+    }
+    
+    /// Attempts to disconnect from all of the relays.
+    public func disconnect() {
+        relays.forEach { $0.disconnect() }
+    }
+    
+    /// Sends a request to the relays.
+    /// - Parameter request: The request to send
+    public func send(request: String) {
+        relays.forEach { $0.send(request: request) }
+    }
+    
+    /// Subscribes to all of the relays using the supplied ``Filter``.
+    /// - Parameters:
+    ///   - filter: The filter to subscribe to.
+    ///   - subscriptionId: The subscription id. If you do not supply one, a random one will be created and returned.
+    /// - Returns: The subscription id.
+    public func subscribe(with filter: Filter, subscriptionId: String = UUID().uuidString) throws -> String {
+        do {
+            try relays.forEach { try $0.subscribe(with: filter, subscriptionId: subscriptionId) }
+        } catch {
```

Comment:
This would fail upon the first error, right? Should we do this or should we make a best effort attempt for all the relays?

Reply:
Yeah. Thinking about how to solve this.

Reply:
Okay, I made this update, and I think it's probably best. However it makes It so that this function no longer `throws`, so the caller won't be notified of errors. It'll still produce messages in the console though, and they're better now because they contain the relay URL.

Reply:
That's true. We can iterate on this over time. Maybe it could make sense to have two approaches exposed. It might still be valuable to bubble up errors in a first-class way while still attempting a best-effort subscription on the relays, and let applications decide how they want to deal with errors.

This is fine for now, though. We can always iterate on it.